src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java

```diff
@@ -145,73 +172,104 @@ private Long getDownloadSizeWithoutRetry() throws InterruptedException, PackageD
-                                            .putHeader(HTTP_RANGE_HEADER_KEY,
-                                                    String.format(HTTP_RANGE_HEADER_FORMAT, rangeStart, rangeEnd))
-                                            .build())
-                            .build();
-                    HttpExecuteResponse executeResponse = client.prepareRequest(executeRequest).call();
+            return RetryUtils.runWithRetry(
+                    clientExceptionRetryConfig,
+                    () -> {
+                        try (SdkHttpClient client = getSdkHttpClient()) {
+                            HttpExecuteRequest executeRequest =
+                                    HttpExecuteRequest.builder()
+                                            .request(
+                                                    SdkHttpFullRequest.builder()
+                                                            .uri(URI.create(url))
+                                                            .method(SdkHttpMethod.GET)
+                                                            .putHeader(
+                                                                    HTTP_RANGE_HEADER_KEY,
+                                                                    String.format(
+                                                                            HTTP_RANGE_HEADER_FORMAT,
+                                                                            rangeStart,
+                                                                            rangeEnd))
+                                                            .build())
+                                            .build();
+                            HttpExecuteResponse executeResponse =
+                                    client.prepareRequest(executeRequest).call();
 
-                    int responseCode = executeResponse.httpResponse().statusCode();
+                            int responseCode = executeResponse.httpResponse().statusCode();
 
-                    // check response code
-                    if (responseCode == HttpURLConnection.HTTP_PARTIAL) {
-                        try (InputStream inputStream = executeResponse.responseBody().get()) {
-                            long downloaded = download(inputStream, messageDigest);
-                            if (downloaded == 0) {
-                                // If 0 byte is read, it's fairly certain that the input stream is closed.
-                                // Therefore throw IOException to trigger the retry logic.
-                                throw new IOException(getErrorString("Failed to read any byte from the stream"));
+                            // check response code
+                            if (responseCode == HttpURLConnection.HTTP_PARTIAL) {
+                                try (InputStream inputStream =
+                                        executeResponse.responseBody().get()) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that you are using `Optional` value which can hold either a value or not. The value held in the `Optional` can be accessed using the `get()` method, but it will throw a `NoSuchElementException` if there is no value present. To avoid the exception, calling the `isPresent()` or `ifPresent()` or `!isEmpty()` method should always be done before any call to `get()`. Alternatively, other methods such as `orElse(...)`, `orElseGet(...)` or `orElseThrow(...)` can be used to specify what to do with an empty `Optional`.

[CWE-476](https://cwe.mitre.org/data/definitions/476)